Isomorphism tutorial (#83)
networkx/nx-guides: content/algorithms/isomorphism/isomorphism.md

```{code-cell}
nx.could_be_isomorphic(G, H)
```

Again we can detect that G and H are not isomorphic. But these conditions are not enough to say that two graphs are isomorphic. Let's look at the following example:

```{code-cell}
plt.subplot(221)
G = nx.path_graph(5)
G.add_edge(2, 5)
nx.draw_circular(G, with_labels=True, node_color="g")
plt.title("G", fontweight="bold")

plt.subplot(222)
H = nx.path_graph(5)
H.add_edge(3, 5)
nx.draw_circular(H, with_labels=True, node_color="c")
plt.title("H", fontweight="bold")
plt.show()
```

```{code-cell}
nx.could_be_isomorphic(G, H)
```

These graphs meet all the necessary conditions but they're not isomorphic.

## Some classes of graphs with solution in polynomial time
- Trees
- Planar graphs(In fact, planar graph isomorphism is O(log(n)))
- Interval graphs
- Permutation graphs
- Circulant graphs
- Bounded-parameter graphs
    - Graphs of bounded treewidth
    - Graphs of bounded genus
    - Graphs of bounded degree
    - Graphs with bounded eigenvalue multiplicity
    - k-Contractible graphs (a generalization of bounded degree and bounded genus)

> Let's see an example, we can use the function *tree_isomorphism()* from the isomorphism module to check if two trees are isomorphic in $O(n*log(n))$. This function uses a D&C approach to match the trees once it has found the root of each tree and returns a list with the node matching. 

Reviewer: I've never heard of `D&C approach` - we should probably more explicitly state what this means!

---
jupytext:
  main_language: python
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.14.2
kernelspec:
  display_name: Python 3 (ipykernel)
  language: python
  name: python3
---

# Isomorphism

```{code-cell}
import networkx as nx
import matplotlib.pyplot as plt
```

## What's isomorphism? Why is it interesting?

As unlabeled graphs can have multiple spatial representations, two graphs are isomorphic if they have the same number of edges, vertices, and same edges connectivity. Let's see an example of two isomorphic graphs,

```{code-cell}
plt.subplot(221)
G = nx.Graph(
    [
        (0, 1),
        (0, 4),
        (0, 2),
        (1, 3),
        (1, 5),
        (2, 3),
        (2, 6),
        (3, 7),
        (4, 5),
        (5, 7),
        (6, 7),
        (6, 4),
    ]
)
nx.draw_spectral(G, with_labels=True, node_color="c")
plt.title("G", fontweight="bold")

> H = nx.Graph(
>     [
>         (1, 2),
>         (1, 5),
>         (1, 3),
>         (2, 4),
>         (2, 6),
>         (3, 4),
>         (3, 7),
>         (4, 8),
>         (5, 6),
>         (6, 8),
>         (7, 8),
>         (7, 5),
>     ]
> )

Reviewer 2: ```suggestion
H = nx.cubical_graph()
```

plt.subplot(222)
nx.draw_circular(H, with_labels=True, node_color="yellow")
plt.title("H", fontweight="bold")

plt.show()
```

> These graphs' spatial representations are very different yet they are isomorphic.

Reviewer 2: ```suggestion
These graphs' spatial representations are very different yet they are the same graphs!
```

### **Formal definition**

G and H are isomorphic if we can establish a bijection between the vertex sets of G and H. 

$$ {\displaystyle f\colon N(G)\to N(H)}$$

such as if 

$v$  and $ w $ are  adjacent  in G $\iff$ $f(v)$ and $f(w)$ are adjacent in H 

To formally prove that 2 graphs are isomorphic we need to find the bijection between the vertex set. For the previous example that would be: 

$$f(i) = i+1 \hspace{0.5cm} \forall i \in [0, 7]$$

For small examples, isomorphism may seem easy. But it isn't a simple problem. For two graphs G and H of n nodes, there are n! bijections function possible. Checking every combination is not a feasible option for bigger graphs. 
In fact, isomorphism is part of the problems known as NP. This means that we don't know any algorithm that runs in polynomial time.

### Applications 

> 
> - Verification of equivalence of different representations of the design of an electronic circuit and communication networks. 
> - Image recognition.
> - Identification of chemical compounds and proteins.
> - Algorithms for fingerprint, facial and retina matching. 
> -  Clustering Algorithms on social networks.
> 
> 

Reviewer 2: ```suggestion

There are a lot of applications of the graph isomorphism problem.

- Image recognition: Images can be translated to graphs and by finding (sub)isomorphisms we can compare if two images are similar.

- Verification of equivalence of different representations of the design of an electronic circuit and communication networks. 

- Identification of chemical compounds and proteins.

- Algorithms for fingerprint, facial and retina matching.

- Clustering Algorithms on social networks.

```
